### PR TITLE
Fixed length epochs overlap

### LINKED ISSUE #8773: Adding overlap keyword to make_fixed_length_epochs
#### Describe the new feature or enhancement
I am trying to use time-frequency analysis that is present in mne without converting to arrays first. There is no way to do this using the ``Raw`` object, and one must convert to ``Epochs``, with each epoch "representing" the window that will have tfr methods applied. E.g. I want to compute the multitaper FFT, and/or morlet wavelet transform to get an ``AverageTfr`` object for multiple frequency bands, alpha, beta, gamma, etc.

Normally, one would like some overlap of these windows. For example, if you do multitpaer FFT, then you usually use a sliding window with some overlap (e.g. 1 sec window with 0.5 sec step).

In https://mne.tools/stable/generated/mne.make_fixed_length_events.html#mne.make_fixed_length_events
there is an overlap keyword.

However, in `make_fixed_legnth_epochs` there is not. 
https://mne.tools/stable/generated/mne.make_fixed_length_epochs.html#mne.make_fixed_length_epochs

#### Describe your proposed implementation
Is it possible to add the `overlap` keyword argument here?

#### Describe possible alternatives
I'm still not quite sure what the optimal way to apply tfr analysis to a Raw object. Obviously one can always convert to arrays, but then they lose the structure present in the mne data structures. I would prefer staying with the data structures mne has (e.g. Epochs for TFR). If what I said up there is not the optimal way to construct TFR then please let me know.

#### Additional comments
I can update an existing example to show this functionality off if needed.

### PR DESCRIPTION
**Pull request by sfc-neuro**
#### Reference issue
Fixes #8773.

#### What does this implement/fix?
Adds compatibility for the overlap parameter in the function make_fixed_length_epochs

#### Additional information
Tests were also added to test the new feature inspired by the pre-existing ones written for make_fixed_length_events.

### CODE REVIEW

File: mne/epochs.py [hunk 1]
```diff
@@ -3667,7 +3667,7 @@ def average_movements(epochs, head_pos=None, orig_sfreq=None, picks=None,
 
 
 @verbose
-def make_fixed_length_epochs(raw, duration=1., preload=False,
+def make_fixed_length_epochs(raw, duration=1., overlap=0., preload=False,
```
Comment (agramfort) [inline mne/epochs.py:3670]: @sfc-neuro you might break the code of people by inserting a new param in the middle. We usually but it just before verbose.

File: mne/epochs.py [hunk 2]
```diff
@@ -3678,6 +3678,9 @@ def make_fixed_length_epochs(raw, duration=1., preload=False,
         Raw data to divide into segments.
     duration : float
         Duration of each epoch in seconds. Defaults to 1.
+    overlap : float
+        The overlap between epochs. Must be ``0 <= overlap < duration``.
+        Default is 0.
```
Comment (agramfort) [inline mne/epochs.py:3683]: you need to add the `versionadded` directive to document when this parameter appeared in a release.

File: mne/viz/ica.py [hunk 3]
```diff
@@ -44,11 +44,11 @@ def plot_ica_sources(ica, inst, picks=None, start=None,
...
-    start : int
-        X-axis start index. If None, from the beginning.
-    stop : int
-        X-axis stop index. If None, next 20 are shown, in case of evoked to the
-        end.
+    start : int | None
+        X-axis start index. If None (default), from the beginning.
+    stop : int | None
+        X-axis stop index. If None (default), next 20 are shown, in case of
+        evoked to the end.
```
Comment (agramfort) [inline mne/viz/ica.py:51]: these commits are unrelated. You should have started a branch from main rather than from your preview PR.

File: mne/epochs.py (hunk 2 above)
Comment (sfc-neuro) [reply mne/epochs.py:3683]: Thanks, added

File: mne/epochs.py (hunk 1 above)
Comment (sfc-neuro) [reply mne/epochs.py:3670]: Thanks, moved.

File: mne/epochs.py [hunk 4]
```diff
@@ -3678,6 +3678,11 @@ def make_fixed_length_epochs(raw, duration=1., preload=False,
...
+
+        .. versionadded:: 0.23
```
Comment (agramfort) [inline mne/epochs.py:3685]: docstring entries must match the order of the parameters. So it should be moved just before verbose

File: mne/viz/ica.py (hunk 3 above)
Comment (sfc-neuro) [reply mne/viz/ica.py]: I think it should be fixed in the next pull request! Thanks for pointing it out

File: mne/epochs.py (hunk 4 above)
Comment (sfc-neuro) [reply mne/epochs.py]: Thanks!

File: mne/tests/test_epochs.py [hunk 5]
```diff
@@ -3254,6 +3254,22 @@ def test_make_fixed_length_epochs():
...
     assert len(epochs) > len(epochs_annot)
 
+    # overlaps
+    epochs = make_fixed_length_epochs(raw, duration=1, preload=True)
+    assert len(epochs) > 10
+    epochs_ol = make_fixed_length_epochs(raw, duration=1, overlap=0.5,
+                                         preload=True)
+    assert len(epochs_ol) > 20
+    epochs_ol_2 = make_fixed_length_epochs(raw, duration=1, overlap=0.9,
+                                           preload=True)
```
Comment (agramfort) [inline mne/tests/test_epochs.py:3264]:
```suggestion
    epochs_ol_2 = make_fixed_length_epochs(raw, duration=1, overlap=0.9)
```

File: mne/tests/test_epochs.py [hunk 6]
```diff
@@ -3254,6 +3254,22 @@ def test_make_fixed_length_epochs():
     assert len(epochs_annot) > 10
     assert len(epochs) > len(epochs_annot)
 
+    # overlaps
+    epochs = make_fixed_length_epochs(raw, duration=1, preload=True)
+    assert len(epochs) > 10
+    epochs_ol = make_fixed_length_epochs(raw, duration=1, overlap=0.5,
+                                         preload=True)
+    assert len(epochs_ol) > 20
```
Comment (agramfort) [inline mne/tests/test_epochs.py:3262]:
```suggestion
    epochs_ol = make_fixed_length_epochs(raw, duration=1, overlap=0.5)
    assert len(epochs_ol) > 20
```

File: mne/tests/test_epochs.py [hunk 7]
```diff
@@ -3254,6 +3254,22 @@ def test_make_fixed_length_epochs():
     assert len(epochs_annot) > 10
     assert len(epochs) > len(epochs_annot)
 
+    # overlaps
+    epochs = make_fixed_length_epochs(raw, duration=1, preload=True)
```
Comment (agramfort) [inline mne/tests/test_epochs.py:3258]:
```suggestion
    epochs = make_fixed_length_epochs(raw, duration=1)
```

Comment (sfc-neuro) [reply mne/tests/test_epochs.py:3258]: I'll comment here bc it is the same for the suggestions before. If I do not specify the preload, the `assert len(epochs)` fails because the len(epochs) seem to only be working with preloaded data.

I thought I should keep it for this reason, what do you think? Could I change the check in a more optimized way?

Comment (sfc-neuro) [reply mne/tests/test_epochs.py]: Edit: I changed it to check the length of the events in the epochs objects to avoid preloading but still performing an analogous check.

Now the tests are not anymore on the same style of the previous ones in the same file (`mne/tests/test_epochs.py`)  that create epochs with `preload=True` and asserting `len(epochs)> n`.

@agramfort do you think they should also be modified too for better readability and optimization?

**Review by agramfort — APPROVED** [review]
can we just add a what's new entry?

thx @sfc-neuro !

**Comment by agramfort** [comment]
@sfc-neuro can you quickly push a what's new entry so we can move on here? thanks

File: doc/changes/latest.inc [hunk 8]
```diff
@@ -39,8 +39,12 @@ Current (0.23.dev0)
 
 .. |Matteo Anelli| replace:: **Matteo Anelli**
 
+.. |Silvia Cotroneo| replace:: **Silvia Cotroneo**
+
 Enhancements
 ~~~~~~~~~~~~
+- Add parameter ``overlap=0.`` to :func:`mne.epochs.make_fixed_length_epochs` to allow creating overlapping fixed length epochs (:gh:`9096` **by new contributor** |Silvia Cotroneo|_)
```
Comment (hoechenberger) [inline doc/changes/latest.inc:46]:
```suggestion
- Add ``overlap`` parameter to :func:`mne.epochs.make_fixed_length_epochs` to allow creating overlapping fixed length epochs (:gh:`9096` **by new contributor** |Silvia Cotroneo|_)
```

File: mne/epochs.py [hunk 9]
```diff
@@ -3685,6 +3685,11 @@ def make_fixed_length_epochs(raw, duration=1., preload=False,
     %(proj_epochs)s
 
         .. versionadded:: 0.22.0
+    overlap : float
+        The overlap between epochs. Must be ``0 <= overlap < duration``.
+        Default is 0.
```
Comment (hoechenberger) [inline mne/epochs.py:3690]:
```suggestion
        The overlap between epochs, in seconds. Must be
        ``0 <= overlap < duration``. Default is 0, i.e., no overlap.
```

**Review by hoechenberger — APPROVED** [review]
LGTM once CI is green

**Comment by wmvanvliet** [comment]
Thanks @sfc-neuro!
